### ISSUE TITLE
[RFC] Add pre/post hooks around Trainer forward_backward_step

### ISSUE THREAD

**Issue by jojoinfra** [pytorch/torchtitan#4026]
# RFC: Add pre/post hooks around Trainer forward_backward_step
## Summary

This RFC proposes adding stable pre/post hook extension points around each call to `Trainer.forward_backward_step()`.

The hooks would allow out-of-tree integrations to observe the boundary of each gradient-accumulation microbatch without:

- overriding or copying `Trainer.train_step()`;
- overriding `Trainer.forward_backward_step()`;
- monkey-patching the global `Trainer` class;
- changing the forward/backward return value.

Example use cases include:

- per-microbatch gradient diagnostics;
- pre-synchronization local gradient norm collection;
- custom loss and gradient monitoring;
- tracing and profiling;
- backend-specific debugging and observability;
- temporary fault-injection and validation tooling.

The proposed hooks are observational and opt-in. When no hooks are registered, existing Trainer behavior remains unchanged.

## Motivation

`Trainer.train_step()` currently invokes `self.forward_backward_step(...)` once for each outer gradient-accumulation microbatch.

Out-of-tree integrations may need to perform lifecycle operations immediately before and after this call. For example:

```text
begin_microbatch()
    -> forward_backward_step()
    -> gradients become available
end_microbatch()
```

There is currently no stable extension point for this boundary. Integrations therefore need to use one of the following approaches:

1. override `Trainer.forward_backward_step()`;
2. override or copy `Trainer.train_step()`;
3. monkey-patch the global `Trainer` class;
4. infer the lifecycle from lower-level model or autograd hooks.

These approaches have several limitations:

- They are tightly coupled to Trainer implementation details.
- Copies of `train_step()` can diverge from upstream behavior.
- Global monkey patches can affect unrelated Trainer instances.
- Multiple independently developed integrations cannot compose safely.
- Subclasses overriding `forward_backward_step()` may bypass wrappers installed by another subclass.
- Exception cleanup is difficult without a post-hook contract.
- Model and autograd hooks do not represent the same lifecycle as an outer gradient-accumulation microbatch.

A small hook API at the Trainer boundary would provide a stable and composable extension point.

## Goals

The proposed API should:

- invoke a pre-hook immediately before each `forward_backward_step()` call;
- invoke a post-hook after the call completes;
- support cleanup when forward/backward raises an exception;
- cover subclasses that override `forward_backward_step()` but inherit the base `train_step()`;
- preserve hook registration order;
- allow hooks to be removed;
- introduce negligible overhead when no hooks are registered;
- avoid requiring hook callables to be represented in job configuration;
- keep existing Trainer behavior unchanged when unused.

## Non-goals

This RFC does not propose:

- replacing PyTorch module forward/backward hooks;
- adding parameter-level or gradient-level hooks;
- exposing individual pipeline schedule microbatches;
- adding an FSDP pre-reduce-scatter gradient API;
- allowing hooks to replace input tensors, loss values, or gradients;
- serializing hook callables into TorchTitan configuration;
- defining backend-specific diagnostic metrics in the core Trainer.

For example, collecting a local gradient norm before FSDP reduce-scatter still requires an FSDP-level observation mechanism. The proposed Trainer hooks only provide the correct outer microbatch lifecycle around that observer.

### Hook registration

Add per-Trainer hook registration methods:

```python
from torch.utils.hooks import RemovableHandle


def register_forward_backward_pre_hook(
    self,
    hook: ForwardBackwardPreHook,
) -> RemovableHandle:
    ...


def register_forward_backward_post_hook(
    self,
    hook: ForwardBackwardPostHook,
) -> RemovableHandle:
    ...
```

A caller registers hooks after constructing the Trainer and before calling `trainer.train()`:

```python
pre_handle = trainer.register_forward_backward_pre_hook(pre_hook)
post_handle = trainer.register_forward_backward_post_hook(post_hook)

trainer.train()

pre_handle.remove()
post_handle.remove()
```

Hooks are stored on each Trainer instance. Registering a hook does not modify the global Trainer class or affect other Trainer instances.

### Proposed hook signatures

One possible initial signature is:

```python
from collections.abc import Callable
from typing import TypeAlias

import torch


ForwardBackwardPreHook: TypeAlias = Callable[
    [
        "Trainer",
        dict[str, torch.Tensor],
        torch.Tensor,
        float | torch.Tensor,
    ],
    None,
]


ForwardBackwardPostHook: TypeAlias = Callable[
    [
        "Trainer",
        torch.Tensor | None,
        BaseException | None,
    ],
    None,
]
```

The arguments have the following meanings:

```text
pre_hook(
    trainer,
    input_dict,
    labels,
    global_valid_tokens,
)

post_hook(
    trainer,
    loss,
    error,
)
```

For a successful forward/backward call:

```text
loss  = the loss returned by forward_backward_step()
error = None
```

For a failed forward/backward call:

```text
loss  = None
error = the original exception
```

The initial API should require hooks to return `None`. Hooks should not replace Trainer inputs or outputs.

### Hook invocation point

The hooks should be invoked by `Trainer.train_step()` around the dynamic dispatch to `self.forward_backward_step(...)`:

```python
self._call_forward_backward_pre_hooks(
    input_dict=input_dict,
    labels=labels,
    global_valid_tokens=global_valid_tokens,
)

loss = None
error = None

try:
    loss = self.forward_backward_step(
        input_dict,
        labels,
        global_valid_tokens,
    )
except BaseException as exc:
    error = exc
    raise
finally:
    self._call_forward_backward_post_hooks(
        loss=loss,
        error=error,
    )
```

This is intentionally placed around the call in `train_step()`, rather than inside the base implementation of `forward_backward_step()`.

As a result:

```text
Trainer.train_step()
    |
    +-- pre hooks
    |
    +-- self.forward_backward_step(...)
    |       |
    |       +-- base implementation
    |       |
    |       `-- or a subclass override
    |
    `-- post hooks
```

A subclass overriding `forward_backward_step()` remains observable as long as it inherits the base `train_step()`.

The pseudocode above omits the final exception precedence policy. The implementation should ensure that a post-hook failure during exception cleanup does not silently replace the original forward/backward exception.

### Microbatch semantics

One hook pair corresponds to one call to `Trainer.forward_backward_step()`.

For gradient accumulation:

```text
optimizer step
    |
    +-- pre hook  [accumulation microbatch 0]
    +-- forward_backward_step()
    +-- post hook [accumulation microbatch 0]
    |
    +-- pre hook  [accumulation microbatch 1]
    +-- forward_backward_step()
    +-- post hook [accumulation microbatch 1]
    |
    `-- optimizer step
```

The hook API does not assign a microbatch index. An integration that requires an index can maintain it in its own hook state.

### Pipeline parallel semantics

For pipeline parallel training, `forward_backward_step()` may invoke a pipeline schedule that internally processes multiple pipeline microbatches.

The proposed hook granularity remains:

```text
one pre/post pair per outer forward_backward_step() call
```

It does not provide hooks for each pipeline-schedule-internal microbatch.

This distinction should be documented explicitly so that integrations do not interpret the outer hook index as a pipeline microbatch index.

A separate pipeline schedule hook API would be needed if pipeline-internal observability is required.

### Hook ordering and removal

Hooks should:

- execute in registration order;
- be registered per Trainer instance;
- return a `RemovableHandle`;
- safely support removal before training or between steps;
- use a fast empty-hook path when no hooks are installed.

A possible internal representation is:

```python
self._forward_backward_pre_hooks = OrderedDict()
self._forward_backward_post_hooks = OrderedDict()
```

This follows existing PyTorch hook conventions without exposing the internal dictionaries as public API.

### Exception behavior

The post-hook is useful for both successful completion and cleanup:

```python
def post_hook(trainer, loss, error):
    if error is not None:
        monitor.abort_microbatch()
        return

    monitor.end_microbatch(loss)
```

The following behavior should be defined:

- The post-hook runs when forward/backward succeeds.
- The post-hook also runs when forward/backward raises.
- The original forward/backward exception remains the primary exception.
- Hook exceptions should include enough context to identify the failing hook.
- All hooks either run in registration order, or execution stops at the first hook failure; this should be selected and documented consistently.

## Example: per-microbatch gradient diagnostics

An out-of-tree integration could use the hooks as follows:

```python
class GradientDiagnosticHooks:
    def __init__(self, monitor):
        self.monitor = monitor
        self.microbatch = 0

    def pre_hook(
        self,
        trainer,
        input_dict,
        labels,
        global_valid_tokens,
    ):
        self.monitor.begin_microbatch(self.microbatch)

    def post_hook(self, trainer, loss, error):
        if error is not None:
            self.monitor.abort_microbatch()
            return

        self.monitor.end_microbatch(loss)
        self.microbatch += 1
```

Registration:

```python
diagnostics = GradientDiagnosticHooks(monitor)

trainer.register_forward_backward_pre_hook(
    diagnostics.pre_hook,
)
trainer.register_forward_backward_post_hook(
    diagnostics.post_hook,
)
```

For an FSDP pre-synchronization local gradient norm diagnostic, the responsibilities would be separated as follows:

```text
Trainer pre-hook
    -> begins the outer microbatch observation window

FSDP reduce-scatter observer
    -> samples gradients before reduce-scatter

Trainer post-hook
    -> closes the observation window
    -> associates the sampled norm with the microbatch
    -> performs cleanup on failure
```

This removes the need to subclass or monkey-patch Trainer while keeping the FSDP-specific behavior outside the core Trainer.

### Override `forward_backward_step()`

An integration can subclass Trainer and wrap the parent implementation:

```python
def forward_backward_step(self, *args, **kwargs):
    monitor.begin_microbatch()
    try:
        return super().forward_backward_step(*args, **kwargs)
    finally:
        monitor.end_microbatch()
```

This works for a single integration, but it has limitations:

- multiple wrappers require an inheritance ordering agreement;
- another subclass may replace the method without calling `super()`;
- backend integrations must introduce a custom Trainer type;
- Trainer construction must be changed to use that type.

### Override or copy `train_step()`

Copying `train_step()` gives complete control but introduces substantial maintenance cost. Changes to gradient accumulation, loss normalization, optimizer behavior, pipeline handling, or metrics must be manually synchronized with upstream.

### Monkey-patch Trainer

Replacing `Trainer.forward_backward_step` globally is easy to prototype, but:

- it affects every Trainer instance in the process;
- patch ordering becomes significant;
- separate extensions may overwrite each other;
- import order determines behavior;
- it is difficult to reason about the final implementation at runtime.

### Model or autograd hooks

Module and autograd hooks operate at a lower level. They may execute multiple times due to:

- activation checkpoint recomputation;
- pipeline schedules;
- shared modules;
- compiled graph behavior;
- multiple parameter groups.

They do not provide the same lifecycle boundary as one outer `forward_backward_step()` call.

### Context manager API

Another possible API is:

```python
with trainer.forward_backward_context(...):
    loss = self.forward_backward_step(...)
```

This provides natural exception handling, but registering multiple independent context managers is less consistent with existing PyTorch hook APIs. A registration-based hook API is likely easier for out-of-tree integrations to compose.

## Backwards Compatibility

The proposal is opt-in:

- no hooks are installed by default;
- existing Trainer subclasses continue to work;
- existing configurations do not need to change;
- hook callables are registered at runtime rather than serialized;
- forward/backward results and optimizer behavior remain unchanged.

Subclasses that override `forward_backward_step()` but inherit the base `train_step()` are covered.

Subclasses that override `train_step()` would need to invoke the shared hook helpers or use a proposed context-manager helper. This limitation should be documented.

## Performance Considerations

When no hooks are registered, the implementation should use an empty-container fast path:

```python
if self._forward_backward_pre_hooks:
    self._call_forward_backward_pre_hooks(...)
```

The expected disabled-path overhead is one predictable container check before and after each outer forward/backward call.

When hooks are enabled, their synchronization and device overhead are owned by the hook implementation. The Trainer hook API itself should not introduce device synchronization or tensor copies.

## Testing

The implementation should include tests covering:

1. pre/post ordering across multiple gradient-accumulation microbatches;
2. post-hook access to the returned loss;
3. post-hook execution when forward/backward raises;
4. preservation of the original forward/backward exception;
5. a subclass overriding `forward_backward_step()`;
6. hook registration order;
7. removal through `RemovableHandle`;
8. multiple independent hooks on one Trainer;
9. unchanged behavior when no hooks are registered;
10. documented PP behavior at the outer `forward_backward_step()` boundary.

## Open Questions

1. Should the hooks be invoked in `train_step()` around the dynamic call, or inside the base `forward_backward_step()` implementation?

2. Should post-hooks always run on failure, or should the API follow `nn.Module` conventions and expose an `always_call` option?

3. Should the pre-hook receive `input_dict`, `labels`, and `global_valid_tokens`, or should the initial API expose only the Trainer instance?

4. Should the post-hook receive both `loss` and `error`, or should error cleanup use a separate hook?

5. Should hooks be strictly observational, or should a future API allow them to replace inputs or the returned loss?

6. Should the public names use the full form:

   ```text
   register_forward_backward_pre_hook
   register_forward_backward_post_hook
   ```

   or the shorter form:

   ```text
   register_fwd_bwd_pre_hook
   register_fwd_bwd_post_hook
   ```

7. Should TorchTitan provide a shared context-manager/helper so that custom Trainers overriding `train_step()` can preserve the same hook contract?

## Expected Impact

The proposed change is intentionally small in the core Trainer, but it provides a stable lifecycle boundary for observability and backend integrations.

It would allow external projects to add per-microbatch diagnostics without maintaining a Trainer fork, copying the training loop, or installing global monkey patches.

If the overall direction and API shape are acceptable, I would be happy to provide an implementation and tests.

**Comment by tianyu-l** [comment]
Thanks, several comments:

[Comment 1]

GPU tensor logging is on our plate, but we haven't got bandwidth to tackle. cc @felipemello1 @scotts 

In your proposal, would it cause perf overhead? How would you deal with it for production runs?


[Comment 2]

> Subclasses overriding forward_backward_step() may bypass wrappers installed by another subclass.

There're not many subclasses of `Trainer`? It sounds viable to me.


[Comment 3]

The hook design reminds me of https://github.com/Lightning-AI/pytorch-lightning. Is it something you are looking for?

**Comment by jojoinfra** [comment]
Thank you for the comments! Sorry that the original RFC was a little abstract. The concrete problem I am trying to solve is the following.

During one training run, I observed an abnormal global grad norm. The final value showed that something was wrong, but it could not tell me which rank or which gradient-accumulation microbatch produced the abnormal gradient.

With FSDP, after reduce-scatter, the original rank-local gradient contributions are no longer directly attributable. To diagnose the problem, I need:

```text
pre-hook
    -> mark the current accumulation microbatch

FSDP reduce-scatter observer
    -> inspect the local gradient before reduce-scatter
    -> associate it with the rank and microbatch

post-hook
    -> finalize the observation or clean up on failure
```

The FSDP observer is still responsible for inspecting the actual gradient. The proposed Trainer hooks only provide the outer microbatch lifecycle. I am not proposing that Trainer expose or process FSDP-internal gradients.

The planned GPU tensor logging work sounds very relevant. My current understanding is that GPU tensor logging handles how device values are collected and logged, while the hooks provide when an observation starts and ends. If the planned logging work already provides this lifecycle boundary, I would be happy to reuse it.

For performance, normal production runs would not register these diagnostic hooks. The no-hook path should call the existing `forward_backward_step()` directly, without entering the hook or additional exception-handling path.

When diagnostics are enabled, gradient collection is not free, so it should be opt-in and limited to configured diagnostic steps. Intermediate values should remain on device during forward/backward, with no `.item()` or device-to-host synchronization on the critical path. Host conversion and logging would happen after the optimizer step.

I agree that subclassing is viable. My current prototype wraps `forward_backward_step()` in a `MyTrainer` subclass. My concern is mainly that an out-of-tree backend must maintain a custom Trainer type only for this lifecycle boundary, and composing it with another specialized Trainer becomes more difficult.

I also understand that hooks placed in the base `train_step()` would not cover Trainers that override `train_step()`. The scope could therefore be limited to Trainers using the base implementation, or represented by a small shared helper if broader coverage is preferred.

The motivation is similar to Lightning callbacks, but I am not looking for a full callback framework—only a narrow, per-Trainer, observational boundary around one outer `forward_backward_step()` call.

I hope this makes the problem and the intended scope clearer. I would really appreciate your guidance on whether this direction fits TorchTitan.
